### ISSUE TITLE
[BUG][DART] anyOf serializer throws exception

### ISSUE THREAD

**Issue by szekelyisz** [OpenAPITools/openapi-generator#14519]
#### Bug Report Checklist

- [X] Have you provided a full/minimal spec to reproduce the issue?
- [X] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [X] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [X] Have you searched for related issues/PRs?
- [X] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

The generated Dart serializer code for `anyOf` schemas throws an exception when the referenced types are also schemas themselves.

##### openapi-generator version

6.3.0-SNAPSHOT 90e468b9a3bf48bf81dde68014d5452356d906b1

##### OpenAPI declaration file content or url

<details>
<summary>swagger.json</summary>

```json
{
	"components": {
		"examples": {},
		"headers": {},
		"parameters": {},
		"requestBodies": {},
		"responses": {},
		"schemas": {
			"NetworkConfigManual": {
				"properties": {
					"method": {
						"type": "string",
						"enum": [
							"manual"
						],
						"nullable": false
					},
					"address": {
						"type": "string"
					},
					"prefix": {
						"type": "integer",
						"format": "int32"
					},
					"gateway": {
						"type": "string"
					},
					"dnss": {
						"items": {
							"type": "string"
						},
						"type": "array"
					}
				},
				"required": [
					"method",
					"address",
					"prefix",
					"dnss"
				],
				"type": "object",
				"additionalProperties": false
			},
			"NetworkConfigAuto": {
				"properties": {
					"method": {
						"type": "string",
						"enum": [
							"auto"
						],
						"nullable": false
					}
				},
				"required": [
					"method"
				],
				"type": "object",
				"additionalProperties": false
			},
			"NetworkConfig": {
				"anyOf": [
					{
						"$ref": "#/components/schemas/NetworkConfigManual"
					},
					{
						"$ref": "#/components/schemas/NetworkConfigAuto"
					}
				]
			}
		},
		"securitySchemes": {}
	},
	"info": {
		"title": "MyAPI",
		"version": "1.0.0"
	},
	"openapi": "3.0.0",
	"paths": {}
}
```
</details>

##### Generation Details

`generate -i swagger.json -g dart-dio -o my_api/generated --additional-properties=legacyDiscriminatorBehavior=false`

##### Steps to reproduce

Generate Dart code using the above schema and command.

Not sure if it's correct but this how I pass data to this serializer:

```dart
          networkConfig: (NetworkConfigBuilder()
                ..anyOf = AnyOf2(values: {
                  1: (NetworkConfigManualBuilder()
                        ..method = NetworkConfigManualMethodEnum.manual
                        ..address = address.address
                        ..prefix = prefix
                        ..gateway = gateway?.address
                        ..dnss = ListBuilder(dnss.map((e) => e.address)))
                      .build()
                }))
              .build()
```

##### Suggest a fix

Here's the generated Dart serializer code for the `NetworkConfig` schema from the above specification:

```dart
  @override
  Object serialize(
    Serializers serializers,
    NetworkConfig object, {
    FullType specifiedType = FullType.unspecified,
  }) {
    final anyOf = object.anyOf;
    final result = <Object?>[];
    for (var _valueEntry in anyOf.values.entries) {
      final _typeIndex = _valueEntry.key;
      final _type = anyOf.types[_typeIndex];
      final _value = _valueEntry.value;
      result.addAll(serializers.serialize(_value, specifiedType: FullType(_type)) as Iterable<Object?>);
    }
    return result;
  }
```

`serializers` become `BuiltJsonSerializers` in this case. The problem here is that the returned value by `serializers.serialize` is a `_InternalLinkedHashMap<String, Object?>` and it is force-cast to `Iterable<Object?>` which obviously throws an exception.

Maybe there needs to be a distinction whether the referenced data types are schemas or primitives and serialization needs to branch accordingly? I don't have enough insight into openapi-generator to tell this.

**Comment by ahmednfwela** [comment]
@szekelyisz thanks for reporting this, can you please test passing the types to `AnyOf2` ?
so it will be like this:
https://github.com/Bdaya-Dev/dart-oneof/blob/main/one_of_serializer_tests/test/any_of_serializer_tests.dart#L122-L130

**Comment by szekelyisz** [comment]
> @szekelyisz thanks for reporting this, can you please test passing the types to `AnyOf2` ? so it will be like this: https://github.com/Bdaya-Dev/dart-oneof/blob/main/one_of_serializer_tests/test/any_of_serializer_tests.dart#L122-L130

@ahmednfwela  I believe I tried that, and it didn't make any difference since the compiler infers the types correctly, explicitly specifying it makes no difference. And so the serializer still returns a map, an the cast fails the same way.

Later I saw your one_of_serializer package and that it actually has the correct code in it. And it was already registered in BuiltJsonSerializers, so it was there but not being used because the generated code took over it, but with a buggy implementation. So I just delegated the serialization to your package and it started to work. This is what the above PR is about.
